### PR TITLE
Add noorul room storage

### LINKED ISSUE #56: Fixed size array in RoomList
The current implementation of `RoomList` uses a fixed size array. Integration with the UI is difficult because changes to the number of `Room` is not properly reflected.

Instead of an array, perhaps we could use an `ObservableList<Room>` to store `Room` objects.

### LINKED ISSUE #57: RoomOccupancyStorageTest: files remain after completion of tests
Great work with implementing and unit testing storage 👍! Definitely a complex task.

Currently, multiple files are created in the root directory. These files are not deleted even after the tests complete and may cause different behaviors across multiple unit test runs. Perhaps we can set a temporary parent folder to store the files to avoid this issue. For consistency, it may be a good idea to follow `JsonAddressBookStorageTest`. It uses a temporary directory as denoted by `@TempDir`.

### PR DESCRIPTION
**Pull request by itssodium**
Please review my PR where I changed files from .txt to JSON and removed arrays and used ObservableList.

### CODE REVIEW

File: src/main/java/seedu/address/storage/RoomOccupancyStorage.java
Comment (LeeMingDe) [inline src/main/java/seedu/address/storage/RoomOccupancyStorage.java]: if this part is not needed, then will it be better to just delete it?

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (LeeMingDe) [inline src/main/java/seedu/address/model/room/RoomList.java:41]: i think you can delete this line and subsequently, the convertPriorityQueue method

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (LeeMingDe) [inline src/main/java/seedu/address/model/room/RoomList.java:112]: can delete the method below. i dont think it's in use anymore. i only created it to bypass the need for an observablelist

**Review by LeeMingDe — APPROVED** [review]
lgtm, would really appreciate it if you can remove the convertPriorityQueue method for me

File: docs/UserGuide.md
Comment (w-yeehong) [inline docs/UserGuide.md:135]: I think it would be good to mention something along the lines of how the command replaces the current number of rooms with the newly specified value. Also, whether the data in the current rooms will be migrated or reset.

We can always improve on the user guide in the future though, so maybe we can KIV? 😄

File: src/main/java/seedu/address/MainApp.java
Comment (w-yeehong) [inline src/main/java/seedu/address/MainApp.java]: Since there are multiple files being checked, maybe we can include which data file is not in the correct format in the warning.

The same goes for the other warnings.

File: src/main/java/seedu/address/MainApp.java
Comment (w-yeehong) [inline src/main/java/seedu/address/MainApp.java:64]: Maybe we can keep this statement in a single line such that it looks even neater? It has fewer than 110 characters and can fit in one line!

File: src/main/java/seedu/address/commons/util/FileUtil.java
Comment (w-yeehong) [inline src/main/java/seedu/address/commons/util/FileUtil.java]: Perhaps we can remove code that is commented out. If we merge this, we may forget to remove it in the future. Also, commented out code seems like regular comments and may be difficult to locate in a large codebase.

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/room/RoomList.java:165]: Great work with the hashing!

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/room/RoomList.java:166]: May I ask why the 31 is needed here?

File: src/main/java/seedu/address/model/room/ReadOnlyRoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/room/ReadOnlyRoomList.java:6]: I think maybe we can make this more general and apply it all the other *List classes (UniquePatientList, RoomList, TaskList). Otherwise, I think the interface may not be needed?

Possibly `ReadOnlyList<T>`? We can do either:

1. Include all methods such as `add`, `delete`, etc. (much more work because we have to rename all methods to share a common name).
2. Include just the method to obtain a read-only list (easier).

File: src/main/java/seedu/address/model/util/SampleDataUtil.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:44]: Maybe we can add a TODO here to indicate that in the future, we should set the application default to having a few rooms with patients and tasks?

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/room/RoomList.java]: Although `ObservableList` and `PriorityQueue` maintain the same rooms, for safety would it be better to include a comparison for the `ObservableList` too?

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/room/RoomList.java]: The condition for this is `rooms != null && roomList.rooms != null` but I think neither the `PriorityQueue` nor `ObservableList` can be `null` because of the default value, right?

In that case, can you not just return `false` if any of the 2 `PriorityQueue` and 2 `ObservableList` are `null`?

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/model/room/RoomList.java:20]: Instead of tracking `numOfRooms` in a variable, would it be possible to just make use of the size of the `ObservableList`? Maybe include a method to get size (similar to the other *List classes).

This would reduce the complexity of the class, especially when writing the `equals(...)` or `hashCode()` method since there is one fewer variable to manage.

File: src/main/java/seedu/address/storage/JsonAdaptedRoom.java
Comment (w-yeehong) [inline src/main/java/seedu/address/storage/JsonAdaptedRoom.java]: `@JsonProperty` tells the Jackson ObjectMapper to map the JSON property name to the annotated Java field's name so we end up with this:

```
"room": {
    "roomNumber": value
    "isOccupied": value
     //...
```

For consistency with `JsonAdaptedPerson`, perhaps we can keep it the same as the variable name (i.e. roomNumber and isOccupied)? We would have

```
public JsonAdaptedRoom(@JsonProperty("roomNumber") int roomNumber,
        @JsonProperty("isOccupied") boolean isOccupied) {
//...
```

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (itssodium) [reply src/main/java/seedu/address/model/room/RoomList.java:112]: I need it for the Json.

File: src/main/java/seedu/address/storage/JsonAdaptedRoom.java
Comment (w-yeehong) [inline src/main/java/seedu/address/storage/JsonAdaptedRoom.java:15]: We can add a TODO here since we are not using the `Patient` and `Task` in this class yet.

File: src/main/java/seedu/address/storage/RoomOccupancyStorage.java
Comment (itssodium) [reply src/main/java/seedu/address/storage/RoomOccupancyStorage.java]: Okay will do!

File: src/main/java/seedu/address/storage/JsonSerializableRoomList.java
Comment (w-yeehong) [inline src/main/java/seedu/address/storage/JsonSerializableRoomList.java]: Consistent `@JsonProperty` here! Variable name should be `rooms`.

File: src/test/java/seedu/address/logic/LogicManagerTest.java
Comment (w-yeehong) [inline src/test/java/seedu/address/logic/LogicManagerTest.java:53]: Good job with switching to temp folders! 😊

File: src/test/java/seedu/address/storage/RoomOccupancyStorageTest.java
Comment (w-yeehong) [inline src/test/java/seedu/address/storage/RoomOccupancyStorageTest.java]: Maybe we can import java.util.Optional directly?

File: src/test/java/seedu/address/testutil/RoomBuilder.java
Comment (w-yeehong) [inline src/test/java/seedu/address/testutil/RoomBuilder.java:12]: I will add this when I pull your changes 👍

File: src/test/java/seedu/address/testutil/RoomUtil.java
Comment (w-yeehong) [inline src/test/java/seedu/address/testutil/RoomUtil.java]: I think if we are not using this class currently, we can remove it 1st?

**Review by w-yeehong — CHANGES_REQUESTED** [review]
Thank you for your hard work! This is a major change, since the entity (`Room`) we are adding is so complex. I think there are some code quality nits we can discuss and address before we merge? Really good work though. You even went as far as to integrate Jackson for Room even though it's not needed for this part yet.

File: src/main/java/seedu/address/storage/JsonAdaptedRoom.java
Comment (itssodium) [reply src/main/java/seedu/address/storage/JsonAdaptedRoom.java]: Okay

File: docs/UserGuide.md
Comment (chiamyunqing) [inline docs/UserGuide.md:183]: Should be after search patient command but it's okay for now. We will do a massive UG clean up in future as a group :)

File: src/main/java/seedu/address/MainApp.java
Comment (chiamyunqing) [inline src/main/java/seedu/address/MainApp.java:94]: Will be starting with an empty "room list" or something instead of AddressBook. Just to clean up traces of address book from the code haha.

File: src/main/java/seedu/address/model/room/RoomList.java
Comment (chiamyunqing) [reply src/main/java/seedu/address/model/room/RoomList.java:20]: I agree with this too. There may be errors if we forget to update this variable when needed.

**Review by chiamyunqing — APPROVED** [review]
Good job and effort! Just some nitty gritty details. :)

File: src/main/java/seedu/address/MainApp.java
Comment (itssodium) [reply src/main/java/seedu/address/MainApp.java:94]: That was wat I was wondering too, but didn't want to delete before discussing.

**Review by w-yeehong — APPROVED** [review]
LGTM!

**Pull request merged**
